### PR TITLE
Bugfix: Liked screen crashing

### LINKED ISSUE #243: As a user, I would like to use the app in offline mode and see the itineraries i liked previously

### LINKED ISSUE #264: Fix bugs with saving itineraries

### PR DESCRIPTION
**Pull request by ethangraham2001**
# Bug fixes for offline mode. Closes #243 Closes #264 

## Changes
- `Location` nullable fields changed to non-nullable ones. Google protobuf doesn't like null values and was causing serialization problems (crashed app)
- Refactored some function calls, such as `isNetworkAvailable()` which is now an extension function for `Context` class instead of being defined locally for `itineraryRepository`
- Modified some tests for compatibility with changed code
- Added a save button to `LikedScreen` (saves are no longer implicit)
- Added a hint popup to `LikedScreen`
- Fixed how liked itineraries were being fetched offline
- Fetching liked itinerary UIDs relies on firebase caching (requires sign-in to have been successful on app launch)

## `LikedScreen` UI changes (in offline mode)

Screenshot from 2024-05-19 12-57-53

### CODE REVIEW

File: app/src/main/java/com/github/wanderwise_inc/app/data/ItineraryRepositoryImpl.kt
Comment (imaneb942) [inline app/src/main/java/com/github/wanderwise_inc/app/data/ItineraryRepositoryImpl.kt:1]: looks good to me, more modularity achieved which is good

File: app/src/main/java/com/github/wanderwise_inc/app/model/location/Location.kt
Comment (imaneb942) [inline app/src/main/java/com/github/wanderwise_inc/app/model/location/Location.kt:1]: what does this change?

File: app/src/main/java/com/github/wanderwise_inc/app/ui/list_itineraries/ItineraryList.kt
Comment (imaneb942) [inline app/src/main/java/com/github/wanderwise_inc/app/ui/list_itineraries/ItineraryList.kt:1]: the name does look better XD

**Review by imaneb942 — APPROVED** [review]
a few refactoring here and there and safer implementation, this looks good to me

File: app/src/main/java/com/github/wanderwise_inc/app/model/location/Location.kt
Comment (ethangraham2001) [reply app/src/main/java/com/github/wanderwise_inc/app/model/location/Location.kt:1]: protobufs (used for local storage) doesn't like null values. Changed it to empty defaults.

Comment (ethangraham2001) [reply app/src/main/java/com/github/wanderwise_inc/app/model/location/Location.kt:1]: protobufs (used for local storage) doesn't like null values

File: app/src/main/java/com/github/wanderwise_inc/app/ui/list_itineraries/LikedScreen.kt [hunk 1]
```diff
@@ -129,4 +149,10 @@ fun DisplayLikedItineraries(
             parent = ItineraryListParent.LIKED,
             imageRepository = imageRepository)
       }
+  var isHintVisible by remember { mutableStateOf(true) }
+  if (isHintVisible) {
```
[the remark this answers is not in the document]
Comment (ethangraham2001) [reply app/src/main/java/com/github/wanderwise_inc/app/ui/list_itineraries/LikedScreen.kt:153]: It's the icon. It's the little save icon

File: app/src/main/java/com/github/wanderwise_inc/app/data/ItineraryRepositoryImpl.kt
Comment (engjellismaili) [inline app/src/main/java/com/github/wanderwise_inc/app/data/ItineraryRepositoryImpl.kt:1]: Lgtm, good new implementation in offline mode

File: app/src/main/java/com/github/wanderwise_inc/app/ui/list_itineraries/ItineraryList.kt
Comment (engjellismaili) [inline app/src/main/java/com/github/wanderwise_inc/app/ui/list_itineraries/ItineraryList.kt:1]: Good change  the key, looks better

File: app/src/main/java/com/github/wanderwise_inc/app/util.kt
Comment (engjellismaili) [inline app/src/main/java/com/github/wanderwise_inc/app/util.kt:1]: Good new function

**Review by engjellismaili — APPROVED** [review]
Overall lgtm, clean new implementation using offline mode. The fix for the liked screen was a major issue.

**Pull request merged**